collect_app/src/main/java/org/odk/collect/android/widgets/BarcodeWidget.kt

```diff
@@ -0,0 +1,123 @@
+package org.odk.collect.android.widgets
+
+import android.annotation.SuppressLint
+import android.app.Activity
+import android.content.Context
+import android.view.View
+import com.google.zxing.integration.android.IntentIntegrator
+import org.javarosa.core.model.data.IAnswerData
+import org.javarosa.core.model.data.StringData
+import org.javarosa.form.api.FormEntryPrompt
+import org.odk.collect.android.activities.ScannerWithFlashlightActivity
+import org.odk.collect.android.databinding.BarcodeWidgetBinding
+import org.odk.collect.android.formentry.questions.QuestionDetails
+import org.odk.collect.android.utilities.Appearances
+import org.odk.collect.android.utilities.Appearances.hasAppearance
+import org.odk.collect.android.utilities.Appearances.isFrontCameraAppearance
+import org.odk.collect.android.widgets.interfaces.WidgetDataReceiver
+import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry
+import org.odk.collect.androidshared.system.CameraUtils
+import org.odk.collect.androidshared.ui.ToastUtils.showLongToast
+import org.odk.collect.permissions.PermissionListener
+import org.odk.collect.strings.R
+
+@SuppressLint("ViewConstructor")
+class BarcodeWidget(
+    context: Context,
+    questionDetails: QuestionDetails,
+    private val widgetAnswerView: WidgetAnswerView,
+    private val waitingForDataRegistry: WaitingForDataRegistry,
+    private val cameraUtils: CameraUtils,
+    dependencies: Dependencies
+) : QuestionWidget(context, dependencies, questionDetails), WidgetDataReceiver {
+    lateinit var binding: BarcodeWidgetBinding
+
+    init {
+        render()
+    }
+
+    override fun onCreateWidgetView(context: Context, prompt: FormEntryPrompt, answerFontSize: Int): View {
+        binding = BarcodeWidgetBinding.inflate((context as Activity).layoutInflater)
+
+        if (prompt.isReadOnly) {
+            binding.barcodeButton.visibility = GONE
+        } else {
+            binding.barcodeButton.setOnClickListener { onButtonClick() }
+        }
+
+        val answer = prompt.answerText
+        if (!answer.isNullOrEmpty()) {
+            binding.barcodeButton.text = getContext().getString(R.string.replace_barcode)
+        }
+        widgetAnswerView.setAnswer(answer)
+        binding.answerViewContainer.addView(widgetAnswerView)
+        updateAnswerVisibility()
+
+        return binding.root
+    }
+
+    override fun clearAnswer() {
+        widgetAnswerView.setAnswer(null)
+        binding.barcodeButton.text = context.getString(R.string.get_barcode)
+        updateAnswerVisibility()
+        widgetValueChanged()
+    }
+
+    override fun getAnswer(): IAnswerData? {
+        val answer = widgetAnswerView.getAnswer()
```

Comment:
Could we still store state at the widget level here? As I said in the previous comment, removing `getAnswer` from `WidgetAnswerView` so it's not used as a store of state would be ideal here. Ultimately, we'd like to avoid storing state in the widget as well (it should live outside the view layer in a ViewModel or something similar at that level), but for now I'd just like to avoid duplicating that problem in these new components.